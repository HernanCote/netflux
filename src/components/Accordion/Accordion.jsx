/* eslint-disable no-shadow */
import React, {
  createContext,
  useState,
  useContext,
} from 'react';

import {
  Container,
  Inner,
  Title,
  Item,
  Header,
  Body,
} from './StyledComponents';

const ToggleContext = createContext();

const Accordion = ({
  children,
  ...rest
}) => (
  <Container
    {...rest}
  >
    <Inner>{children}</Inner>
  </Container>
);

Accordion.Title = function AccordionTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Accordion.Item = function AccordionItem({ children, ...rest }) {
  const [show, setShow] = useState(false);

  return (
    <ToggleContext.Provider value={{ show, setShow }}>
      <Item {...rest}>{children}</Item>
    </ToggleContext.Provider>
  );
};

Accordion.Header = function AccordionHeader({ children, ...rest }) {
  const { show, setShow } = useContext(ToggleContext);
  return (
    <Header onClick={() => setShow(show => !show)} {...rest}>
      {children}
      {show ? <img src="images/icons/close-slim.png" alt="close" /> : <img src="images/icons/add.png" alt="open" /> }
    </Header>
  );
};

Accordion.Body = function AccordionBody({ children, ...rest }) {
  const { show } = useContext(ToggleContext);
  return show ? <Body {...rest}>{children}</Body> : null;
};

export default Accordion;
