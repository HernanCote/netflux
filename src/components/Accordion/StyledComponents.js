import styled from 'styled-components/macro';

const Container = styled.div`
  display: flex;
  border-bottom: 0.5rem solid #222;
`;

const Inner = styled.div`
  display: flex;
  padding: 4.25rem 2.75rem;
  flex-direction: column;
  max-width: 50rem;
  margin: auto;
`;

const Title = styled.h1`
  font-size: 2.5rem;
  line-height: 1.1rem;
  margin-top: 0;
  margin-bottom: 0.5rem;
  color: #fff;
  text-align: center;

  @media (max-width: 600px) {
    line-height: 2.5rem;
  }
`;

const Item = styled.div`
  color: #fff;
  margin: auto;
  margin-bottom: 0.75rem;
  max-width: 41.75rem;
  width: 100%;

  &:first-of-type {
    margin-top: 3rem;
  }
`;

const Header = styled.div`
  display: flex;
  justify-content: space-between;
  cursor: pointer;
  margin-bottom: 1px;
  font-size: 1.75rem;
  font-weight: normal;
  background: #303030;
  padding: 0.75rem 1.25rem;
  user-select: none;
  align-items: center;

  img {
    filter: brightness(0) invert(1);
    width: 1.5rem;

    @media (max-width: 600px) {
      width: 1rem;
    }
  }

  @media (max-width: 600px) {
    font-size: 1rem;
  }
`;

const Body = styled.div`
  max-height: 75rem;
  transition: max-height 0.25s cubic-bezier(0.5, 0, 0.1, 1);
  font-size: 1.75rem;
  font-weight: normal;
  line-height: normal;
  background: #303030;
  padding: 0.75rem 2.25rem 0.75rem 1.25rem;
  white-space: pre-wrap;
  user-select: none;

  @media (max-width: 600px) {
    font-size: 1rem;
    line-height: 1.25rem;
  }
`;

export {
  Container,
  Inner,
  Title,
  Item,
  Header,
  Body,
};
