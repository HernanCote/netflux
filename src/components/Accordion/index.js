export { default } from './Accordion';
export * from './StyledComponents';
