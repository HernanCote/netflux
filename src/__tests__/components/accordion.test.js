/* eslint-disable no-undef */
/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import faqsData from '../../fixtures/faqs.json';
import Accordion from '../../components/Accordion';

describe('<Accordion />', () => {
  it('renders <Accordion /> with populated data', () => {
    const { container, getByText } = render(
      <Accordion>
        <Accordion.Title>Frequently Asked Questions</Accordion.Title>
        {faqsData.map(item => (
          <Accordion.Item key={item.id}>
            <Accordion.Header>{item.header}</Accordion.Header>
            <Accordion.Body>{item.body}</Accordion.Body>
          </Accordion.Item>
        ))}
      </Accordion>,
    );

    expect(getByText('Frequently Asked Questions')).toBeTruthy();
    expect(getByText(faqsData[0].header)).toBeTruthy();
    expect(getByText(faqsData[1].header)).toBeTruthy();
    expect(getByText(faqsData[2].header)).toBeTruthy();
    expect(getByText(faqsData[3].header)).toBeTruthy();
    expect(getByText(faqsData[4].header)).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('opens and closes the <Accordion /> component', () => {
    const { container, queryByText, queryByTestId } = render(
      <Accordion>
        <Accordion.Title>Frequently Asked Questions</Accordion.Title>
        {faqsData.map(item => (
          <Accordion.Item key={item.id}>
            <Accordion.Header>{item.header}</Accordion.Header>
            <Accordion.Body data-testid={item.id}>{item.body}</Accordion.Body>
          </Accordion.Item>
        ))}
      </Accordion>,
    );

    // eslint-disable-next-line quotes
    const whatIsNetfluxBodyText = faqsData[0].body;
    expect(queryByText(whatIsNetfluxBodyText)).toBeFalsy();
    fireEvent.click(queryByText(faqsData[0].header));
    expect(queryByTestId(1).innerHTML).toContain(whatIsNetfluxBodyText);

    fireEvent.click(queryByText(faqsData[0].header));
    expect(queryByText(whatIsNetfluxBodyText)).toBeFalsy();
    expect(container.firstChild).toMatchSnapshot();
  });
});
