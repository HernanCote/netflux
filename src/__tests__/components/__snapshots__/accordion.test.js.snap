// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> opens and closes the <Accordion /> component 1`] = `
<div
  class="StyledComponents__Container-sc-1gbwx9c-0 gQhybf"
>
  <div
    class="StyledComponents__Inner-sc-1gbwx9c-1 ihmThD"
  >
    <h1
      class="StyledComponents__Title-sc-1gbwx9c-2 daipkN"
    >
      Frequently Asked Questions
    </h1>
    <div
      class="StyledComponents__Item-sc-1gbwx9c-3 cJgWIs"
    >
      <div
        class="StyledComponents__Header-sc-1gbwx9c-4 cNHElk"
      >
        What is Netflux?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="StyledComponents__Item-sc-1gbwx9c-3 cJgWIs"
    >
      <div
        class="StyledComponents__Header-sc-1gbwx9c-4 cNHElk"
      >
        How much does Netflux cost?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="StyledComponents__Item-sc-1gbwx9c-3 cJgWIs"
    >
      <div
        class="StyledComponents__Header-sc-1gbwx9c-4 cNHElk"
      >
        Where can I watch?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="StyledComponents__Item-sc-1gbwx9c-3 cJgWIs"
    >
      <div
        class="StyledComponents__Header-sc-1gbwx9c-4 cNHElk"
      >
        How do I cancel?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="StyledComponents__Item-sc-1gbwx9c-3 cJgWIs"
    >
      <div
        class="StyledComponents__Header-sc-1gbwx9c-4 cNHElk"
      >
        What can I watch on Netflux?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Accordion /> renders <Accordion /> with populated data 1`] = `
<div
  class="StyledComponents__Container-sc-1gbwx9c-0 gQhybf"
>
  <div
    class="StyledComponents__Inner-sc-1gbwx9c-1 ihmThD"
  >
    <h1
      class="StyledComponents__Title-sc-1gbwx9c-2 daipkN"
    >
      Frequently Asked Questions
    </h1>
    <div
      class="StyledComponents__Item-sc-1gbwx9c-3 cJgWIs"
    >
      <div
        class="StyledComponents__Header-sc-1gbwx9c-4 cNHElk"
      >
        What is Netflux?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="StyledComponents__Item-sc-1gbwx9c-3 cJgWIs"
    >
      <div
        class="StyledComponents__Header-sc-1gbwx9c-4 cNHElk"
      >
        How much does Netflux cost?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="StyledComponents__Item-sc-1gbwx9c-3 cJgWIs"
    >
      <div
        class="StyledComponents__Header-sc-1gbwx9c-4 cNHElk"
      >
        Where can I watch?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="StyledComponents__Item-sc-1gbwx9c-3 cJgWIs"
    >
      <div
        class="StyledComponents__Header-sc-1gbwx9c-4 cNHElk"
      >
        How do I cancel?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="StyledComponents__Item-sc-1gbwx9c-3 cJgWIs"
    >
      <div
        class="StyledComponents__Header-sc-1gbwx9c-4 cNHElk"
      >
        What can I watch on Netflux?
        <img
          alt="open"
          src="images/icons/add.png"
        />
      </div>
    </div>
  </div>
</div>
`;
